storage/aptosdb/src/state_store/mod.rs

```diff
@@ -639,36 +658,41 @@ impl StateStore {
         let base_version = first_version.checked_sub(1);
         let mut usage = self.get_usage(base_version)?;
         let base_version_usage = usage;
-        let cache = Arc::new(DashMap::<StateKey, (Version, Option<StateValue>)>::new());
 
+        let mut state_cache_with_version = &arr![DashMap::new(); 16];
         if let Some(base_version) = base_version {
             let _timer = OTHER_TIMERS_SECONDS
                 .with_label_values(&["put_stats_and_indices__total_get"])
                 .start_timer();
-            let key_set = value_state_sets
-                .iter()
-                .flat_map(|value_state_set| value_state_set.iter())
-                .map(|(key, _)| key)
-                .collect::<HashSet<_>>();
-            IO_POOL.scope(|s| {
-                for key in key_set {
-                    let cache = cache.clone();
-                    s.spawn(move |_| {
-                        let _timer = OTHER_TIMERS_SECONDS
-                            .with_label_values(&["put_stats_and_indices__get_state_value"])
-                            .start_timer();
-                        let version_and_value = self
-                            .state_db
-                            .get_state_value_with_version_by_version(key, base_version)
-                            .expect("Must succeed.");
-                        if let Some((version, value)) = version_and_value {
-                            cache.insert(key.clone(), (version, Some(value)));
-                        } else {
-                            cache.insert(key.clone(), (base_version, None));
-                        }
-                    });
-                }
-            });
+            if let Some(sharded_state_cache) = sharded_state_cache {
+                self.prepare_version_in_cache(base_version, sharded_state_cache)?;
```

Comment:
I think this is a proper position to document why/when the versions on the entries can be None.

```
For some entries the base value version is None, here is to fiil those in. See `ShardedStateCache`.
```

Reply:
Done